### PR TITLE
Implement Reshape reference operation

### PR DESCRIPTION
**Pull request by pelszkow**
### Details:
 - *reimplement reference::reshape*
 - *add element number check in visitor API test*

### Tickets:
 - *45955*
 - *53981*

### CODE REVIEW

File: ngraph/core/reference/src/runtime/reference/reshape.cpp [hunk 1]
```diff
@@ -2,46 +2,78 @@
...
+    {
+        std::vector<int64_t> offset(s.size(), 1);
+        for (size_t i = s.size() - 1; i != 0; --i)
+        {
+            offset[i - 1] = offset[i] * s[i];
+        };
+        return offset;
+    }
+
+    void reshape_impl(const char* const arg,
```
Comment (ilyachur) [inline ngraph/core/reference/src/runtime/reference/reshape.cpp:41]: Why did you add `reshape_impl` function? Looks like you can just rewrite the body of `runtime::reference::reshape` function 😄

Comment (pelszkow) [reply ngraph/core/reference/src/runtime/reference/reshape.cpp:41]: Looks like encapsulation freak. I'll move impl to `reshape` body.

**Comment by pelszkow** [comment]
@jdanieck, @ggalieroc, @sdurawa, @ilyachur 
PR is green now and ready for your review.

File: ngraph/test/reshape_opt_kernel.cpp [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+using namespace ngraph;
+
+namespace
+{
+    using ElementValue = int32_t;
+    enum class AxisOrder
+    {
+        straight,
+        reverse,
```
Comment (ilyachur) [inline ngraph/test/reshape_opt_kernel.cpp:24]: Why do we need the reverse order?

File: ngraph/test/reshape_opt_kernel.cpp [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+                                 axis_order,
+                                 p.output.get_shape(),
+                                 sizeof(ElementValue));
+    EXPECT_EQ(p.output.get_vector(), output_buff);
+}
+
+INSTANTIATE_TEST_CASE_P(reshape_opt_kernel,
+                        ReshapeOptKernel,
+                        ::testing::Values(TestParams{AxisOrder::straight,
+                                                     test::NDArray<ElementValue, 2>{
```
Comment (ilyachur) [inline ngraph/test/reshape_opt_kernel.cpp:70]: Do we cover all specific cases? Like without reorder or other root cases

**Comment by mitruska** [comment]
Could you unexclude some reshape backend tests from the [/ie/unit_test.manifest](https://github.com/openvinotoolkit/openvino/blob/master/ngraph/test/runtime/ie/unit_test.manifest#L410-L425)?

File: ngraph/test/reshape_opt_kernel.cpp (hunk 2 above)
Comment (pelszkow) [reply ngraph/test/reshape_opt_kernel.cpp:24]: This allow to parametrize test for two axis order:
straight: {1,2,3,...,n-1}
reverse{n-1,n-2,n-3,..0}

File: ngraph/test/reshape_opt_kernel.cpp (hunk 3 above)
Comment (pelszkow) [reply ngraph/test/reshape_opt_kernel.cpp:70]: Test cover without reorder case `AxisOrder::straight` and reverse order `AxisOrder::reverse` now.

I'm wonder if `AxisOrder` have to validate in ref impl if it has unique values or is this done by ngraph.
Till now other than 'no reorder' and 'reverse order' is not cover.

**Comment by pelszkow** [comment]
@ilyachur, PR is green can you check and merge it?

**Review by ilyachur — APPROVED** [review]

**Comment by pelszkow** [comment]
> Could you unexclude some reshape backend tests from the [/ie/unit_test.manifest](https://github.com/openvinotoolkit/openvino/blob/master/ngraph/test/runtime/ie/unit_test.manifest#L410-L425)?

address in https://github.com/openvinotoolkit/openvino/pull/6161
